tools/type_extract/include/CppEmitter.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <filesystem>
+#include <fstream>
+#include <sstream>
+#include <vector>
+
+#include "slang/util/SmallVector.h"
+#include "slang/util/Util.h"
+
+namespace fs = std::filesystem;
+
+class HppFile {
+public:
+    explicit HppFile(std::string_view name, bool noSystemC) : fileName(std::string(name) + ".h") {
+        includes.emplace_back("ostream");
+        includes.emplace_back("cstddef");
+        includes.emplace_back("cstdint");
+        includes.emplace_back("string");
+        includes.emplace_back("sstream");
+        if (!noSystemC)
+            includes.emplace_back("systemc.h");
+    }
+
+    inline void add(std::string&& code) { hpp << code; }
+    inline void addInclude(std::string&& code) {
+        if (std::find(includes.begin(), includes.end(), code) == includes.end())
+            includes.emplace_back(code);
+    }
+    inline void addIncludeHeader(std::string_view code) {
+        if (std::find(headers.begin(), headers.end(), code) == headers.end())
+            headers.emplace_back(code);
+    }
+    inline void addWithIndent(std::string&& code) { hpp << indent(currentIndent) << code; }
+    inline void increaseIndent() { currentIndent++; }
+    inline void decreaseIndent() {
+        SLANG_ASSERT(currentIndent != 0);
+        currentIndent--;
+    }
+
+    std::string emit() {
+        std::stringstream ss;
+        ss << "// " << fileName << std::endl;
```

Comment:
It's fine if you want to keep what you have but the fmt library can make this kind of code a lot more straightforward.